feat: 카테고리 테이블 생성 및 퀴즈,게임 테이블 수정

youjungHwang opened the pull request:
## #️⃣ 관련 이슈
- #86

## 📝 작업한 내용
- [x] 조회 쿼리에서 과도한 조인 연산 문제를 해결하기 위해 테이블 분리를 선택했습니다.
  - [x] 카테고리 테이블을 추가하고 데이터를 삽입했습니다.
  - [x] 퀴즈 테이블에서 카테고리 컬럼을 삭제하고 카테고리 아이디를 추가했습니다.
  - [x] 게임 테이블에 카테고리 아이디를 추가했습니다.
- [x] FlywayConfig 클래스를 추가하여 마이그레이션 전략을 정의했습니다.
- [x] 데이터베이스 스키마 변경에 따라 기존의 category: String을 categoryId: Long으로 수정했습니다.

## 💬 논의하고 싶은 내용
- Flyway가 적용된 SQL이 먼저 실행되어 퀴즈 테이블에 카테고리 컬럼이 없는 상태에서 기존 init_insert_quiz_data.sql이 정상적으로 실행되지 못하는 이슈가 발생했습니다. 이에 따라 추가 파일을 생성하여 처리하지 않고, 기존의 init_insert_quiz_data.sql 내에서 카테고리 컬럼을 카테고리 아이디로 변경했습니다. 더 좋은 방법이 있었을지 논의해보고 싶습니다. 😃😃

youjungHwang committed:
feat: 카테고리 테이블 생성 및 퀴즈,게임 테이블 수정

- V1__init.sql 파일에서 코드 정리를 수행했습니다.
- 카테고리 테이블을 추가하고 데이터를 삽입했습니다.
- 퀴즈 테이블에서 카테고리 컬럼을 삭제하고 카테고리 아이디를 추가했습니다.
- 게임 테이블에 카테고리 아이디를 추가했습니다.
- flyway가 적용된 SQL이 먼저 실행되므로 카테고리 컬럼이 없는 상태입니다. 이에 따라 init_insert_quiz_data.sql 내에서 카테고리 컬럼을 카테고리 아이디로 직접 변경했습니다.

youjungHwang committed:
feat: flyway 설정 파일 추가

- FlywayConfig 클래스를 추가하여 마이그레이션 전략을 정의했습니다.
- 마이그레이션 실행 전에 flyway.repair()로 체크섬을 업데이트하고, 이후 flyway.migrate()로 실제 마이그레이션을 수행합니다.

youjungHwang committed:
feat: category 도메인 추가

youjungHwang committed:
refactor: category를 categoryId로 변경

- 데이터베이스 스키마 변경에 따라 기존의 category: String을 categoryId: Long으로 수정했습니다.

youjungHwang committed:
test: category를 categoryId로 변경

Jeongjjuna commented on src/main/kotlin/yjh/cstar/category/domain/CategoryType.kt:
```diff
@@ -0,0 +1,19 @@
+import yjh.cstar.common.BaseErrorCode
+import yjh.cstar.common.BaseException
+
+enum class CategoryType(val id: Long, val description: String) {
+    ALGORITHM(1L, "알고리즘"),
+    DATABASE(2L, "데이터베이스"),
+    DESIGN_PATTERN(3L, "디자인패턴"),
+    NETWORK(4L, "네트워크"),
+    OPERATING_SYSTEM(5L, "운영체제"),
+    ;
```
카테고리타입의 첫번째 인자로 id를 가지는 이유가 궁금합니다!

youjungHwang replied:
최대한 간단하게 가져가려고 도메인만 만들고 진행했었습니다. 하지만 카테고리 테이블이 있기에 매핑할 카테고리 엔티티를 생성해야할 것같습니다. 그럼 해당 클래스에서는 아이디를 제거하는 게 맞는 것 같습니다!

Jeongjjuna commented on src/main/kotlin/yjh/cstar/config/FlywayConfig.kt:
```diff
@@ -0,0 +1,16 @@
+@Configuration
+class FlywayConfig {
+    @Bean
+    fun cleanMigrateStrategy(): FlywayMigrationStrategy {
```
혹시 어떤 마이그레이션 설정인지 여쭤봐도 될까요?!

youjungHwang replied:
V3__create_category_table_and_alter_quiz_game_tables.sql을 생성하고 애플리케이션 실행 했을 때, checksum 버전이 일치하지 않는 이슈가 있었습니다. 그래서 FlywayConfig파일을 생성하고 마이그레이션이 실행되기 전에 flyway.repair()를 호출하여 체크섬을 업데이트하고, 이후 flyway.migrate()를 호출하여 실제 마이그레이션을 수행하는 로직을 추가하게 되었습니다!
해당 과정은 노션에 '알게된 학습 정보들' 페이지에도 기재했습니다. 😄😄

Jeongjjuna commented on src/main/resources/db/V3__create_category_table_and_alter_quiz_game_tables.sql:
```diff
@@ -0,0 +1,21 @@
+ALTER TABLE game
+    ADD COLUMN category_id BIGINT;
+
+INSERT INTO category (category_id, category)
+VALUES (1, 'ALGORITHM'),
+       (2, 'DATABASE'),
+       (3, 'DESIGN_PATTERN'),
+       (4, 'NETWORK'),
+       (5, 'OPERATING_SYSTEM');
```
데이터를 넣는 로직은 data폴터 파일에 옮겨줘야 될 것 같습니다!!

youjungHwang replied:
아! 네 바로 수정해서 올리겠습니다!

youjungHwang committed:
refactor: category 엔티티 생성

youjungHwang committed:
refactor: DDL을 data.sql로 이동

youjungHwang committed:
test: 테스트 시 도메인 id 대신 엔티티 id를 사용

Jeongjjuna committed:
test: 실패한 테스트를 성공하도록 수정

Jeongjjuna committed:
style: ktlint 적용

youjungHwang committed:
refactor: Flyway 설정 파일 삭제

- V3 파일 추가로 인한 체크섬 이슈가 아닌 Flyway가 적용된 V1 파일의 코드 정렬로 인해 체크섬 문제가 발생했을 것으로 추측됩니다.
- 따라서 Flyway 마이그레이션 전략 파일은 불필요하다고 판단해 삭제했습니다.
- 이미 적용된 파일 수정은 지양해야합니다.

Jeongjjuna reviewed (approved):
넵 고생하셨어요!!
테스트 관련해서는 제대로 동작방식 확인해보면 좋을 것 같습니다!!
